cfgrammar/src/lib/header.rs

```diff
@@ -0,0 +1,326 @@
+                        kind: HeaderErrorKind::ExpectedToken('}'),
+                        spans: vec![Span::new(section_start_pos, self.src.len())],
+                    });
+                    Err(errs)
+                }
+            } else {
+                errs.push(HeaderError {
+                    kind: HeaderErrorKind::ExpectedToken('{'),
+                    spans: vec![Span::new(i, i)],
+                });
+                Err(errs)
+            }
+        } else if self.required {
+            errs.push(HeaderError {
+                kind: HeaderErrorKind::MissingGrmtoolsSection,
+                spans: vec![Span::new(0, 0)],
+            });
+            Err(errs)
+        } else {
+            Ok((HashMap::new(), 0))
+        }
+    }
+
+    fn parse_name(&self, i: usize) -> Result<(&str, usize), HeaderError> {
+        match RE_NAME.find(&self.src[i..]) {
+            Some(m) => {
+                assert_eq!(m.start(), 0);
+                Ok((&self.src[i..i + m.end()], i + m.end()))
+            }
+            None => Err(HeaderError {
+                kind: HeaderErrorKind::IllegalName,
+                spans: vec![Span::new(i, i)],
+            }),
+        }
+    }
+
+    fn lookahead_is(&self, s: &'static str, i: usize) -> Option<usize> {
+        if self.src[i..].starts_with(s) {
+            Some(i + s.len())
+        } else {
+            None
+        }
+    }
+
+    fn parse_ws(&self, i: usize) -> usize {
+        RE_LEADING_WS
+            .find(&self.src[i..])
+            .map(|m| m.end() + i)
+            .unwrap_or(i)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_header_missing_curly_bracket() {
+        let src = "%grmtools { a, b";
+        for flag in [true, false] {
+            let parser = GrmtoolsSectionParser::new(src, flag);
+            let res = parser.parse();
+            assert!(res.is_err());
+        }
+    }
+
+    #[test]
+    fn test_header_missing_curly_bracket_empty() {
+        let src = "%grmtools {";
+        for flag in [true, false] {
+            let parser = GrmtoolsSectionParser::new(src, flag);
+            let res = parser.parse();
+            assert!(res.is_err());
+        }
+    }
+
+    #[test]
+    fn test_header_missing_curly_bracket_invalid() {
+        let src = "%grmtools {####";
+        for flag in [true, false] {
+            let parser = GrmtoolsSectionParser::new(src, flag);
+            let res = parser.parse();
+            assert!(res.is_err());
+        }
+    }
+
+    #[test]
+    fn test_header_duplicates() {
+        let src = "%grmtools {dupe, !dupe, dupe: test}";
+        for flag in [true, false] {
+            let parser = GrmtoolsSectionParser::new(src, flag);
+            let res = parser.parse();
+            let errs = res.unwrap_err();
+            assert_eq!(errs.len(), 1);
+            assert_eq!(errs[0].kind, HeaderErrorKind::DuplicateEntry);
+            assert_eq!(errs[0].spans.len(), 3);
+
```

Comment:
newline..

Reply:
Squashed a rustfmt in [104b442](https://github.com/softdevteam/grmtools/pull/544/commits/104b44248421631504da66263eea7ffb04bbbcbe)